### PR TITLE
Options rework

### LINKED ISSUE #19: Follow-up: add validateParam helper
https://github.com/NebulousLabs/skynet-js/pull/385#discussion_r604895309

### PR DESCRIPTION
**Pull request by m-cat**
Commit 1:
- [BREAKING] Removed getEntry 'timeout' option.
- Added input validation functions.
- Moved base options to 'utils/options.ts' file.

Commit 2:
- Rework generic 'endpoint' option to be specific for every endpoint (e.g. endpointGetEntry).

Closes #19

### CODE REVIEW

File: src/utils/validation.ts [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+
+/**
+ * Validates the given value as a string.
+ *
+ * @param name - The name of the value.
+ * @param value - The actual value.
+ * @param valueKind - The kind of value that is being checked (e.g. "parameter", "response field", etc.)
+ * @throws - Will throw if not a valid string.
+ */
+export function validateString(name: string, value: unknown, valueKind: string): void {
```
Comment (peterjan) [inline src/utils/validation.ts:85]: maybe we could use convenience method `validateParam(name: string, value: unknown)` seeing as 85% of all cases are `validateString("paramname", paramvalue, "parameter")`

File: src/utils/validation.ts [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+  valueKind: string,
+  model: Record<string, unknown>
+): void {
+  if (!value) {
+    // This is okay, the object is optional.
+    return;
+  }
+  validateObject(name, value, valueKind);
+
+  // Check if all given properties of value also exist in the model.
```
Comment (peterjan) [inline src/utils/validation.ts:55]: not sure if I'd be thrilled about this as an SDK user

File: src/upload.ts [hunk 3]
```diff
@@ -195,3 +227,25 @@ export async function uploadDirectoryRequest(
...
+  }
+}
+
+function validateUploadResponse(response: AxiosResponse): void {
+  try {
+    if (!response.data) {
+      throw new Error("response.data field missing");
+    }
+
+    validateString("response.data.skylink", response.data.skylink, "upload response field");
```
Comment (peterjan) [inline src/upload.ts:243]: not sure these 3 instances warranted the abstraction where you need to pass `parameter` or `upload reponse field` - but I don't mind I think.

I wouldn't mention `response.data` in the field name though. Isn't that weird? I think `response.data` is kind of implied by `upload response field`.

**Review by peterjan — APPROVED** [review]

File: src/download.ts [hunk 4]
```diff
@@ -77,16 +79,44 @@ export type ResolveHnsResponse = {
 };
 
 export const defaultDownloadOptions = {
-  ...defaultOptions("/"),
+  ...defaultBaseOptions,
+  endpointDownload: "/",
```
Comment (peterjan) [inline src/download.ts:83]: why do you have to explicitly add these?

File: src/download.ts [hunk 5]
```diff
@@ -77,16 +79,44 @@ export type ResolveHnsResponse = {
...
   hnsSubdomain: "hns",
 };
 const defaultResolveHnsOptions = {
-  ...defaultOptions("/hnsres"),
+  ...defaultBaseOptions,
+  endpointDownloadHnsres: "hnsres",
 };
 
+/**
+ * Extract only the download custom options from the given options.
```
Comment (peterjan) [inline src/download.ts:101]:
```suggestion
 * Extract only the custom download options from the given options.
```

File: src/download.ts [hunk 6]
```diff
@@ -77,16 +79,44 @@ export type ResolveHnsResponse = {
...
+  const downloadOpts = (({ endpointDownload, download, noResponseMetadata, path, query, subdomain }) => ({
+    endpointDownload,
+    download,
+    noResponseMetadata,
+    path,
+    query,
+    subdomain,
+  }))(opts);
+  // @ts-expect-error - We can't ensure the correct types here.
+  return { ...baseOpts, ...downloadOpts };
```
Comment (peterjan) [inline src/download.ts:117]: couldn't you do this using basic destructuring? 
If you have an object like this:
```
const opts = {
    "name": "foo",
    "path": "bar",
    "size": "baz",
    "date": "hello",
}
```
you can destructure a subset by doing:
```
const { name, path, ...rest } = opts;
const subset = {name, path};
return subset, rest; 
```
^ very much pseudo code but the idea should be clear

File: src/utils/validation.ts [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+/**
+ * Validates the given value as an optional object.
+ *
+ * @param name - The name of the value.
+ * @param value - The actual value.
+ * @param valueKind - The kind of value that is being checked (e.g. "parameter", "response field", etc.)
+ * @param model - A model object that contains all possible fields. 'value' does not need to have all fields, but it may not have any fields not contained in 'model'.
+ * @throws - Will throw if not a valid optional object.
+ */
+export function validateOptionalObject(
```
Comment (peterjan) [inline src/utils/validation.ts:43]: think it makes more sense to pass `optional` as a bool to `validateObject`

File: src/download.ts (hunk 4 above)
Comment (m-cat) [reply src/download.ts:83]: the SkyDB options are a combination of registry, download and upload options. This means if someone set the generic `endpointPath` option he would set it for all operations in SkyDB. I needed to make method-specific endpoint options.

File: src/download.ts (hunk 6 above)
Comment (m-cat) [reply src/download.ts]: That's basically what this did :P Not relevant anymore however.

File: src/utils/validation.ts (hunk 1 above)
Comment (m-cat) [reply src/utils/validation.ts:85]: Well there's different input types apart from just "string" and we'd need a convenience method for all of them :P I do agree it's pretty repetetive, idk

**Comment by m-cat** [comment]
Merging without review since this is blocking me a bit from more urgent stuff, and this PR is mostly internal. Reviewers can feel free to leave follow-up issues if they spot anything amiss.

**Pull request merged**